kernel/sched.c

```diff
@@ -1189,24 +1196,18 @@ void z_impl_k_wakeup(k_tid_t thread)
 {
 	SYS_PORT_TRACING_OBJ_FUNC(k_thread, wakeup, thread);
 
-	if (z_is_thread_pending(thread)) {
-		return;
-	}
+	k_spinlock_key_t  key = k_spin_lock(&_sched_spinlock);
 
-	if (z_abort_thread_timeout(thread) < 0) {
-		/* Might have just been sleeping forever */
-		if (thread->base.thread_state != _THREAD_SUSPENDED) {
-			return;
-		}
+	if (!z_is_thread_sleeping(thread)) {
+		k_spin_unlock(&_sched_spinlock, key);
+		return;
 	}
 
-	k_spinlock_key_t  key = k_spin_lock(&_sched_spinlock);
+	(void)z_abort_thread_timeout(thread);
```

Comment:
Great catch! I forgot to double check about that ... will follow up on that.

Reply:
I've been pondering this more and doing more reading of the code trying to keep an eye open for races.

As far as I can tell, we do have cases where  both _sched_spinlock and timeout_lock are both locked, and it seems OK as long as _sched_spinlock was locked first and unlocked last.

On SMP, I think it is better to put the call to z_abort_thread_timeout() in k_wakeup() within the _sched_spinlock is locked section. It may still be OK to have it outside this section, but I find the logic to prove it gets more complicated (and easy to make a mistake in).

If the call to z_abort_thread_timeout() precedes the locking of _sched_spinlock, consider the following ...

 1. CPU1 is executing thread1 and CPU2 is executing thread2.
 2. CPU1 calls k_wakeup(thread2) and CPU2 calls k_sleep(thread2).
 3. CPU1 aborts thread2's non-existent timeout.
 4. CPU1 and CPU2 start competing for _sched_spinlock, but CPU2 wins.
 5. CPU2 finishes k_sleep() for thread2 (creating a new timeout for it) and unlocks _sched_spinlock.
 6. CPU1 gets _sched_spinlock() and finishes its call to k_wakeup(thread2); however due to the order of things, thread2's timeout is not aborted.
 
This leads to the question "What happens when thread2's timeout expires and it was not sleeping?"
   - I *think* it's ok, but proving it gets complicated and error prone.
 That complexity can be bypassed by putting the call to z_abort_thread_timeout() within k_wakeup()'s _sched_spinlock locked section.